Good-Bids/goodbids
# [#374] Stripe Webhooks

# Summary

This PR implements Stripe Webhooks for Invoice-related events, including:
* `invoice.sent` Marks an Invoice as Sent
* `invoice.paid` Marks an Invoice as Paid and stores the Payment ID.
* `invoice.overdue` - This webhook currently doesn't do anything.

## Issues

* #374 

## Testing Instructions

`invoice.paid` is the only Webhook supported by local development. Once deployed to `develop`, we can test `invoice.sent`.
1. From the Main GoodBids site, go to WooCommerce > Settings > Payments > Stripe > Manage > Settings > Edit Account Keys.
2. Follow the instructions in order to retrieve a "TEST WEBHOOK SECRET" (Locally, this will require installing the Stripe CLI)
3. Save the settings with the Webhook Secret.

### Testing Locally
1. Run the following command in your terminal after setting up the Stripe CLI: `stripe listen --forward-to "URL_OF_YOUR_LOCAL_SITE" --skip-verify` - This listens for webhooks and forwards them to your local site.
2. In a separate terminal window, run the following command: `stripe trigger invoice.paid`

Stripe will generate a new customer, invoice, invoice item, payment, and finalize the invoice. You should be able to see this data in your `/private/logs/` directory. This will simulate the Webhook event and if it shows up in the logs, it's working correctly.

### Testing on Develop (in theory)
1. Wait for a new Auction to end.
2. If an invoice is not generated automatically, visit the Auction on the front-end to have it generated.
3. The invoice should get created, which should trigger a Webhook when it is sent.
4. Visiting Nonprofit WP Admin > Invoices you should see the invoice with a status of either "Unsent" or "Unpaid".
5. Click the Pay Now button and use a test CC to pay the invoice.
6. Return to the Invoices page (reload if necessary)
7. The Invoice Status should show "Paid" and the Payment column should show the date paid.

## Screenshots

<img width="1085" alt="Screenshot 2024-02-08 at 3 30 08 PM" src="https://github.com/Good-Bids/goodbids/assets/122309362/62c94513-5cac-40fc-b913-23388d5b93e8">

<img width="1101" alt="Screenshot 2024-02-08 at 3 48 29 PM" src="https://github.com/Good-Bids/goodbids/assets/122309362/d87e36a2-8f89-4a6d-801c-be3fe4d87409">
